File: typescript/test/processor/order-sequential.test.ts
import { mkdtempSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

import { describe, expect, it } from 'vitest';

import { TDC } from '../../src/lib/tdc.js';

const NOW = new Date('2026-04-23T12:00:00Z').getTime();

function lines(out: string): string[] {
  return out.split('\n').filter((l) => l.length > 0);
}

describe('order="sequential" — text', () => {
  const cfg = (count: number, extra = ''): string =>
    `<tdc><env count="${String(count)}" seed="demo"><sequence name="M">` +
    `<gen type="text" value="Jan,Feb,Mar" order="sequential"${extra}/></sequence></env>` +
    '<block><line><data>${{M}}</data></line></block></tdc>';

  it('emits values in order and loops (both engines agree)', () => {
    const expected = ['Jan', 'Feb', 'Mar', 'Jan', 'Feb', 'Mar', 'Jan'];
    expect(lines(new TDC({ configString: cfg(7), now: NOW }).toString())).toEqual(expected);
    expect(new TDC({ configString: cfg(7), now: NOW }).toArray().map((r) => r['M'])).toEqual(
      expected,
    );
  });

  it('cycle="false" errors once the data runs out', () => {
    expect(() => new TDC({ configString: cfg(5, ' cycle="false"'), now: NOW }).toString()).toThrow(
      /only 3 values/,
    );
    // exactly N rows is fine
    expect(lines(new TDC({ configString: cfg(3, ' cycle="false"'), now: NOW }).toString())).toEqual(
      ['Jan', 'Feb', 'Mar'],
    );
  });
});

describe('order="sequential" — file', () => {
  it('replays file lines in their exact order, looping', () => {
    const dir = mkdtempSync(join(tmpdir(), 'tdc-seq-'));
    const path = join(dir, 'villages.txt');
    writeFileSync(path, 'McMurdo\nVostok\nConcordia\n', 'utf8');
    const cfg =
      '<tdc><env count="5" seed="demo"><sequence name="V">' +
      `<gen type="file" src="${path}" order="sequential"/></sequence></env>` +
      '<block><line><data>${{V}}</data></line></block></tdc>';
    const expected = ['McMurdo', 'Vostok', 'Concordia', 'McMurdo', 'Vostok'];
    expect(lines(new TDC({ configString: cfg, now: NOW }).toString())).toEqual(expected);
    expect(new TDC({ configString: cfg, now: NOW }).toArray().map((r) => r['V'])).toEqual(expected);
  });
});

File: java/src/test/java/io/github/nickliapin/tdc/OrderSequentialTest.java
package io.github.nickliapin.tdc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * {@code order="sequential"} against a file, and the refusal when the data runs out.
 *
 * <p>The looping and no-draw behaviour is pinned by the shared cases. These two cannot be: a
 * shared case carries no side files, and an error is not an output.
 */
class OrderSequentialTest {

  @Test
  @DisplayName("a file is replayed in its exact order, looping")
  void fileInOrder(@TempDir Path dir) throws IOException {
    // From typescript/test/processor/order-sequential.test.ts.
    Path file = dir.resolve("villages.txt");
    Files.writeString(file, "McMurdo\nVostok\nConcordia\n");

    TDC tdc =
        TDC.options()
            .configString(
                "<tdc><env mode=\"memory\" count=\"5\" seed=\"demo\" local=\"en\">"
                    + "<sequence name=\"V\"><gen type=\"file\" src=\"villages.txt\" order=\"sequential\"/></sequence>"
                    + "</env><block><line><data>${{V}}</data></line></block></tdc>")
            .baseDir(dir)
            .build();

    assertEquals(
        List.of("McMurdo", "Vostok", "Concordia", "McMurdo", "Vostok"),
        tdc.toString().lines().toList());
    // The rows agree with the text, since both read one run.
    assertEquals(
        List.of("McMurdo", "Vostok", "Concordia", "McMurdo", "Vostok"),
        tdc.toList().stream().map(r -> r.get("V")).toList());
  }

  @Test
  @DisplayName("cycle=\"false\" reports how far the data went")
  void refusesToLoopWhenTold() {
    IllegalStateException e =
        assertThrows(
            IllegalStateException.class,
            () ->
                TDC.options()
                    .configString(
                        "<tdc><env mode=\"memory\" count=\"5\" seed=\"demo\" local=\"en\">"
                            + "<sequence name=\"M\"><gen type=\"text\" value=\"Jan,Feb,Mar\""
                            + " order=\"sequential\" cycle=\"false\"/></sequence>"
                            + "</env><block><line><data>${{M}}</data></line></block></tdc>")
                    .build()
                    .toString());
    // The reference's wording, and its arithmetic: three values, and the fourth row is where it
    // ran out.
    assertTrue(e.getMessage().contains("only 3 values, so row 4 has none"), e.getMessage());
  }
}
